### ISSUE TITLE
List Item not re-render when renderItem changes

### ISSUE THREAD

**Issue by mk-nickyang** [Shopify/flash-list#699]
## Current behavior

List Item does not re-render when `renderItem` prop changes.

## Expected behavior

If this is a bug, List Item should re-render when `renderItem` prop changes,
If this is expected, I think it should be mentioned in the docs since it's different from the RN `FlatList` behaviour.

## To Reproduce

https://snack.expo.dev/@lucissa/shopify-flashlist-renderitem
(`renderItem` gets updated when the `time` state updates, but List Item still shows the initial time)

## Platform:

- [x] iOS
- [x] Android

## Environment

1.4.0

**Comment by mk-nickyang** [comment]
Yeah, I guess there are a couple of ways to trigger re-render manually, like using `extraData` (from https://github.com/Shopify/flash-list/issues/585#issuecomment-1250020456), or using `keyExtractor` like you said, or even just change the `data` prop.

I'm more just want to confirm if this is an expected behaviour due to the FlashList optimization logic. And as I said in the issue above, I think it would be a good idea to mention this in the docs.

**Comment by naqvitalha** [comment]
We'd really appreciate a PR updating the documentation :)

**Comment by hotaryuzaki** [comment]
we found similar issue if we have a checkbox in the list, the checkbox not re-render when we checked.
i think it will be heavy if we change all key every checked checkbox.
so we use extraData and give them all the objects data that checked.

**Comment by matheusmaske** [comment]
@hotaryuzaki your answer help-me a lot, thanks!

**Comment by andresfmm** [comment]
@hotaryuzaki thanks for your answer i leave a example here base in your answer with checkbox


const [extradata, setExtraData] = useState(null);

const [first, setFirst] = useState([{"checked": true, "code_postal": "760013", "id": "22943e1e-f6be-11ed-8207-9828a60f7a9e", "name": "one"}, {"checked": false, "code_postal": "760042", "id": "229444b3-f6be-11ed-8207-9828a60f7a9e", "name": "two"}])


 const keyExtractor = (item:any) => item.id;

const onItemSelected = (item:any) => {


    item.checked = !item.checked;

    for (let index in first) {
        if (first[index].id === item.id) {
            setExtraData(item.id)
        }
    }
    
  }

const elementRender = ({item}:any) => {

     return (
      <TouchableOpacity
        onPress={() => onItemSelected(item)} 
        activeOpacity={0.5}>
        <Checkbox.Item label={item.name} status={ item.checked ? 'checked' : 'unchecked' } />
      
    </TouchableOpacity>
     )
 }

<FlashList
      data={data}
      extraData={extradata}
      keyExtractor={keyExtractor} 
      renderItem={ elementRender }
      estimatedItemSize={700}
 />

**Comment by andresfmm** [comment]
@hotaryuzaki thanks for your answer i leave a example here base in your answer with checkbox

const [extradata, setExtraData] = useState(null);

const [data, setData] = useState([{"checked": true, "code_postal": "760013", "id": "22943e1e-f6be-11ed-8207-9828a60f7a9e", "name": "one"}, {"checked": false, "code_postal": "760042", "id": "229444b3-f6be-11ed-8207-9828a60f7a9e", "name": "two"}])

const keyExtractor = (item:any) => item.id;

const onItemSelected = (item:any) => {

item.checked = !item.checked;

for (let index in data) {
    if (data[index].id === item.id) {
        data[index] = item;
        setExtraData(item.id)
    }
}
}

const elementRender = ({item}:any) => {

 return (
  <TouchableOpacity
    onPress={() => onItemSelected(item)} 
    activeOpacity={0.5}>
    <Checkbox.Item label={item.name} status={ item.checked ? 'checked' : 'unchecked' } />
  
</TouchableOpacity>
 )
}

<FlashList
   data={data}
   extraData={extradata}
   keyExtractor={keyExtractor}
   renderItem={ elementRender }
   estimatedItemSize={700}
/>

**Comment by Hosam-hsm** [comment]
I have a multi select list which marks the selected items. I'm using a state to pass the data prop to the flashlist and I'm storing the selected item in a separate state. An already selected item will get deselected if pressed again. An item gets marked as selected and deselected on first and second press respectively. The list is failing to re-render after that in android even if I am passing the extraData prop. it is working as expected in iOS with the extraData prop.

`const [assets, setAssets] = useState<AssetType[]>([
    {
      id: "1",
      uri: "file:///storage/emulated/0/Pictures/image-7 (1).jpg",
    },
  ]);

  const [selectedMedia, setSelectedMedia] = useState<AssetType[]>([]);

  const onPressMedia = useCallback((media: AssetType) => {
    setSelectedMedia((selected) => {
      const index = selected.findIndex((m) => m.id === media.id);
      if (index !== -1) {
        return [...selected.slice(0, index), ...selected.slice(index + 1)];
      } else {
        return [...selected, media];
      }
    });
  }, []);


  const renderItem = useCallback(
    ({ item, index }) => {
      return (
        <List
          item={item}
          onPressMedia={onPressMedia}
          selectedMedia={selectedMedia}
        />
      );
    },
    [selectedMedia],
  );

and inside return I'm using 
    <FlashList
          estimatedItemSize={height / 6}
          data={assets}
          numColumns={3}
          keyExtractor={(item) => item.id}
          renderItem={renderItem}
          extraData={selectedMedia}
        />
`

**Comment by tyx2000** [comment]
hi there, another issue comes to me, to display a list with FlashList and just want to switch the renderItem in single/double column,
but i change the mode with useState, the list not rerender to another display mode, and FlatList could.

version    expo 51.0.17  FlashList  1.6.4

code is here

```
<FlashList
        data={initData}
        keyExtractor={(item) => item.id.toString()}
        estimatedItemSize={100}
        renderItem={({ item }) => {
          if (item.name === "cate") {
            return (
              <ListMode
                listMode={listMode}
                setListMode={(type: string) => setListMode(type)}
              />
            );
          } else {
            return listMode === "single" ? (
              <SingleColumn item={item} />
            ) : (
              <DoubleColumn item={item} />
            );
          }
        }}
        stickyHeaderIndices={[0]}
      />
```

if is there a solution or just use FlatList and ignore the performance? thanks.

**Comment by audn** [comment]
Can you give this a try? 
I got it working by extracting `extraData` from `renderItem()`.

```tsx
const renderItem = useCallback(
  ({ item, index, extraData }) => {
    return (
      <List item={item} onPressMedia={onPressMedia} selectedMedia={extraData} />
    );
  },
  [selectedMedia]
);

<FlashList
  estimatedItemSize={height / 6}
  data={assets}
  numColumns={3}
  keyExtractor={(item) => item.id}
  renderItem={renderItem}
  extraData={selectedMedia}
/>;

```

cc @Hosam-hsm

**Comment by frozencap** [comment]
> extraData

it works, but why is this even necessary

**Comment by Ajmal0197** [comment]
This worked for me:

```
  const [listData, setListData] = useState([]);

  const setLongDescItemFun = (v) => {
    setLongDescItem(v);
    setListData([...data]);
  };


      <FlashList
        data={data}
        extraData={listData}
        ...
       />

```

**Comment by enzzoperez** [comment]
If you just need to change or test your list ui, you can add `extraData={[]}` to the list and with this could use the hot reload. Then you must delete it

```
<FlashList
        data={data}
        extraData={[]}
        ...
 />

```

**Comment by trajano** [comment]
In my case I added a key extractor that has the same dependencies as the render item.
